### PR TITLE
Application Import Export Support for APIM 2.2.0

### PR DESCRIPTION
**Pull request by randilu**
### Proposed changes in this pull request

- This is the Application Import Export Support for WSO2 API Manager 2.2.0
- This enables users to migrate applications between different environments with or without adding subscriptions as desired by the importing user
- Users are able to export an application as a zip archive file, which in return can be imported to a 
 another environment (dev/QA)
- This also supports import/exports in Multi-Tenanted environments as well.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

- Docs : https://docs.google.com/document/d/1i_tyr-GlloVujhxxADeQTqtBu3rOFHq8aLZJwdkWJ4U

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+                FileUtils.deleteDirectory(new File(path));
+            }
+        } catch (IOException e) {
+            String errorMsg = "Unable to find Application Details at: " + applicationArtifactBaseDirectoryPath;
+            throw new APIManagementException(errorMsg);
+        }
+        return applicationArtifactBaseDirectoryPath;
+    }
+
+    public Application importApplication(InputStream uploadedAppArchiveInputStream) throws APIManagementException {
```
Comment (anuruddhal) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: Add method comments

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+            // cleanup the archive root directory
+            try {
+                FileUtils.deleteDirectory(new File(path));
+            } catch (IOException e1) {
+                log.warn("Unable to remove directory " + path);
+            }
+            String errorMsg = "Error while archiving directory " + sourceDirectory;
+            throw new APIManagementException(errorMsg);
+        }
+        archivedFilePath = archiveLocation + File.separator + archiveName + ".zip";
```
Comment (anuruddhal) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: Shall we move ".zip" extension to a constant?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import org.wso2.carbon.apimgt.rest.api.store.utils.RestAPIStoreUtils;
+import org.wso2.carbon.apimgt.rest.api.util.utils.RestApiUtil;
+import java.io.File;
+import java.util.UUID;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+public class ExportApiServiceImpl extends ExportApiService {
+
+    private static final Log log = LogFactory.getLog(ApplicationsApiServiceImpl.class);
```
Comment (anuruddhal) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImpl.java]: Is the logger class correct?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.io.File;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.UUID;
+import javax.ws.rs.core.Response;
+
+public class ImportApiServiceImpl extends ImportApiService {
+    private static final Log log = LogFactory.getLog(ApplicationsApiServiceImpl.class);
```
Comment (anuruddhal) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java]: Is the logger class correct?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            RestApiUtil
+                    .handleInternalServerError("Error while importing Application" + username, e, log);
+        }
+        return null;
+    }
+
+    @Override
+    public String importApplicationsPostGetLastUpdatedTime(InputStream fileInputStream, Attachment fileDetail,
+                                                           Boolean preserveOwner, Boolean addSubscriptions) {
+        return null;
```
Comment (anuruddhal) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java]: Why is this method not implemented?

Comment (randilu) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java]: this generated method cannot be implemented

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ExportApi.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        
+        @io.swagger.annotations.ApiResponse(code = 404, message = "Not Found.\nRequested Application does not exist.\n"),
+        
+        @io.swagger.annotations.ApiResponse(code = 406, message = "Not Acceptable.\nThe requested media type is not supported\n"),
+        
+        @io.swagger.annotations.ApiResponse(code = 412, message = "Precondition Failed.\nThe request has not been performed because one of the preconditions is not met.\n") })
+
+    public Response exportApplicationsGet(@ApiParam(value = "Application Search Query\n",required=true) @QueryParam("appId")  String appId)
+    {
+    return delegate.exportApplicationsGet(appId);
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ExportApi.java]: Formatting issue.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                UUID.randomUUID().toString(); //creates a directory in default temporary-file directory
+        String username = RestApiUtil.getLoggedInUsername();
+        try {
+            consumer = RestApiUtil.getConsumer(username);
+            FileBasedApplicationImportExportManager importExportManager = new FileBasedApplicationImportExportManager
+                    (consumer, pathToExportDir);
+            applicationDetails = importExportManager.getApplicationDetails(appId);
+            if (applicationDetails == null) {
+                // 404
+                String errorMsg = "No application found for query " + appId;
```
[the remark this answers is not in the document]
Comment (randilu) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImpl.java]: This errorMsg is used again in the response as well

**Pull request closed without merging**